File: packages/rabbitmq/src/consumer.ts
/**
 * RabbitMQ Consumer Implementation
 */

import amqp, { type ChannelModel, type Channel, type ConsumeMessage } from 'amqplib';
import {
  BaseConsumer,
  createMessage,
  type MessageHandler,
  type BatchMessageHandler,
  type SubscribeOptions,
  type HealthStatus,
  ConnectionError,
} from '@anyq/core';
import type { RabbitMQConsumerConfig } from './config.js';
import { RABBITMQ_DEFAULTS } from './config.js';

/**
 * RabbitMQ Consumer
 *
 * Consumes messages from RabbitMQ queues
 */
export class RabbitMQConsumer<T = unknown> extends BaseConsumer<T> {
  private connection: ChannelModel | null = null;
  private channel: Channel | null = null;
  private consumerTag: string | null = null;
  private readonly rabbitConfig: RabbitMQConsumerConfig;

  constructor(config: RabbitMQConsumerConfig) {
    super(config);
    this.rabbitConfig = {
      ...config,
      connection: {
        ...RABBITMQ_DEFAULTS.connection,
        ...config.connection,
      },
      queue: {
        ...RABBITMQ_DEFAULTS.queue,
        ...config.queue,
      },
      consumer: {
        ...RABBITMQ_DEFAULTS.consumer,
        ...config.consumer,
      },
    };
  }

  /**
   * Connect to RabbitMQ
   */
  async connect(): Promise<void> {
    if (this._connected) {
      return;
    }

    try {
      const { connection: connConfig, queue: queueConfig, exchange, bindingKey } = this.rabbitConfig;

      // Build connection URL
      let url = connConfig.url;
      if (connConfig.vhost) {
        url = `${url}/${encodeURIComponent(connConfig.vhost)}`;
      }

      // Connect
      this.connection = await amqp.connect(url, {
        heartbeat: connConfig.heartbeat,
        timeout: connConfig.connectionTimeout,
        channelMax: connConfig.channelMax,
        frameMax: connConfig.frameMax,
      });

      // Create channel
      this.channel = await this.connection.createChannel();

      // Set prefetch
      await this.channel.prefetch(this.rabbitConfig.consumer?.prefetch ?? 10);

      // Assert queue
      const queueArgs: Record<string, unknown> = {};
      if (queueConfig.deadLetterExchange) {
        queueArgs['x-dead-letter-exchange'] = queueConfig.deadLetterExchange;
        if (queueConfig.deadLetterRoutingKey) {
          queueArgs['x-dead-letter-routing-key'] = queueConfig.deadLetterRoutingKey;
        }
      }
      if (queueConfig.messageTtl) {
        queueArgs['x-message-ttl'] = queueConfig.messageTtl;
      }
      if (queueConfig.maxLength) {
        queueArgs['x-max-length'] = queueConfig.maxLength;
      }
      if (queueConfig.maxLengthBytes) {
        queueArgs['x-max-length-bytes'] = queueConfig.maxLengthBytes;
      }

      await this.channel.assertQueue(queueConfig.name, {
        durable: queueConfig.durable,
        exclusive: queueConfig.exclusive,
        autoDelete: queueConfig.autoDelete,
        arguments: { ...queueArgs, ...queueConfig.arguments },
      });

      // If exchange is specified, assert and bind
      if (exchange) {
        await this.channel.assertExchange(exchange.name, exchange.type, {
          durable: exchange.durable ?? true,
          autoDelete: exchange.autoDelete ?? false,
          internal: exchange.internal,
          arguments: exchange.arguments,
        });

        await this.channel.bindQueue(
          queueConfig.name,
          exchange.name,
          bindingKey ?? ''
        );
      }

      // Handle connection errors
      this.connection.on('error', (err) => {
        this.logger.error('Connection error', { error: err.message });
        this._connected = false;
      });

      this.connection.on('close', () => {
        this.logger.warn('Connection closed');
        this._connected = false;
      });

      this._connected = true;
      this.logger.info('RabbitMQ consumer connected', {
        url: connConfig.url,
        queue: queueConfig.name,
        exchange: exchange?.name,
      });
    } catch (error) {
      throw new ConnectionError(
        'Failed to connect to RabbitMQ',
        error instanceof Error ? error : undefined
      );
    }

    // After a successful connect, enforce the park-downgrade policy. RabbitMQ
    // has no native delayed redelivery in this release, so a park-capable
    // strategy downgrades to a blocking in-process retry (warn, or throw when
    // allowParkDowngrade === false). Kept outside the try so a thrown
    // ConfigurationError is not masked as a ConnectionError.
    this.verifyParkPolicy();
  }

  /**
   * Disconnect from RabbitMQ
   */
  async disconnect(): Promise<void> {
    this.beginShutdown();
    try {
      if (this.channel && this.consumerTag) {
        await this.channel.cancel(this.consumerTag);
        this.consumerTag = null;
      }

      if (this.channel) {
        await this.channel.close();
        this.channel = null;
      }

      if (this.connection) {
        await this.connection.close();
        this.connection = null;
      }

      this._connected = false;
      this.logger.info('RabbitMQ consumer disconnected');
    } catch (error) {
      this.logger.error('Error during disconnect', { error });
    }
  }

  /**
   * Subscribe to messages
   */
  async subscribe(
    handler: MessageHandler<T>,
    _options?: SubscribeOptions
  ): Promise<void> {
    if (!this.channel) {
      throw new ConnectionError('Consumer not connected');
    }

    const queueName = this.rabbitConfig.queue.name;
    const consumerOptions = this.rabbitConfig.consumer;

    const result = await this.channel.consume(
      queueName,
      async (msg: ConsumeMessage | null) => {
        if (!msg) {
          return;
        }

        if (this._paused) {
          // Reject and requeue when paused
          this.channel?.nack(msg, false, true);
          return;
        }

        const message = this.convertMessage(msg);
        try {
          await handler(message);
        } catch (error) {
          const err = error instanceof Error ? error : new Error(String(error));
          const result = await this.applyStrategy(message, err, () =>
            handler(message),
          );
          if (!result.handled) {
            this.logger.error('Error processing message', { error });
            // Legacy: message will be nacked by the handler or handled by the adapter
          }
        }
      },
      {
        consumerTag: consumerOptions?.consumerTag,
        noLocal: consumerOptions?.noLocal,
        noAck: consumerOptions?.noAck,
        exclusive: consumerOptions?.exclusive,
        priority: consumerOptions?.priority,
        arguments: consumerOptions?.arguments,
      }
    );

    this.consumerTag = result.consumerTag;

    this.logger.info('Subscribed to queue', {
      queue: queueName,
      consumerTag: this.consumerTag,
    });
  }

  /**
   * Subscribe to messages in batches
   */
  async subscribeBatch(
    handler: BatchMessageHandler<T>,
    options?: SubscribeOptions
  ): Promise<void> {
    if (!this.channel) {
      throw new ConnectionError('Consumer not connected');
    }

    const queueName = this.rabbitConfig.queue.name;
    const consumerOptions = this.rabbitConfig.consumer;
    const batchSize = options?.batchSize ?? 10;
    const batchTimeout = options?.batchTimeout ?? 5000;

    let batch: ConsumeMessage[] = [];
    let batchTimer: ReturnType<typeof setTimeout> | null = null;

    const processBatch = async () => {
      if (batch.length === 0) return;

      const currentBatch = [...batch];
      batch = [];

      if (batchTimer) {
        clearTimeout(batchTimer);
        batchTimer = null;
      }

      const messages = currentBatch.map(msg => this.convertMessage(msg));
      try {
        await handler(messages);
      } catch (error) {
        const err = error instanceof Error ? error : new Error(String(error));
        // RabbitMQ supports per-message ack; apply strategy per message.
        let anyUnhandled = false;
        for (const message of messages) {
          const result = await this.applyStrategy(message, err, () =>
            handler([message]),
          );
          if (!result.handled) {
            anyUnhandled = true;
          }
        }
        if (anyUnhandled) {
          this.logger.error('Error processing batch', { error });
          // Legacy: nack all in-flight messages with requeue.
          for (const msg of currentBatch) {
            this.channel?.nack(msg, false, true);
          }
        }
      }
    };

    const result = await this.channel.consume(
      queueName,
      async (msg: ConsumeMessage | null) => {
        if (!msg) {
          return;
        }

        if (this._paused) {
          this.channel?.nack(msg, false, true);
          return;
        }

        batch.push(msg);

        if (batch.length >= batchSize) {
          await processBatch();
        } else if (!batchTimer) {
          batchTimer = setTimeout(() => processBatch(), batchTimeout);
        }
      },
      {
        consumerTag: consumerOptions?.consumerTag,
        noLocal: consumerOptions?.noLocal,
        noAck: false, // Always manual ack for batch
        exclusive: consumerOptions?.exclusive,
        priority: consumerOptions?.priority,
        arguments: consumerOptions?.arguments,
      }
    );

    this.consumerTag = result.consumerTag;

    this.logger.info('Subscribed to queue (batch mode)', {
      queue: queueName,
      consumerTag: this.consumerTag,
      batchSize,
      batchTimeout,
    });
  }

  /**
   * Convert RabbitMQ message to IMessage format
   */
  private convertMessage(msg: ConsumeMessage) {
    const content = msg.content.toString('utf8');
    let body: T;

    try {
      body = JSON.parse(content) as T;
    } catch {
      body = content as unknown as T;
    }

    // Extract headers
    const headers: Record<string, string | Buffer | undefined> = {};
    if (msg.properties.headers) {
      for (const [key, value] of Object.entries(msg.properties.headers)) {
        if (Buffer.isBuffer(value)) {
          headers[key] = value;
        } else if (value !== undefined) {
          headers[key] = String(value);
        }
      }
    }

    const messageId = msg.properties.messageId ?? `rmq-${msg.fields.deliveryTag}`;
    const exchange = msg.fields.exchange;
    const routingKey = msg.fields.routingKey;
    const deliveryTag = msg.fields.deliveryTag;
    const redelivered = msg.fields.redelivered;

    return createMessage<T>({
      id: messageId,
      body,
      key: routingKey,
      headers,
      timestamp: msg.properties.timestamp
        ? new Date(msg.properties.timestamp)
        : new Date(),
      deliveryAttempt: redelivered ? 2 : 1,
      metadata: {
        provider: 'rabbitmq',
        rabbitmq: {
          exchange,
          routingKey,
          consumerTag: this.consumerTag ?? '',
          deliveryTag,
          redelivered,
        },
      },
      raw: msg,
      onAck: async () => {
        this.channel?.ack(msg);
        this.logger.debug('Message acknowledged', { messageId, deliveryTag });
      },
      onNack: async (requeue = true) => {
        this.channel?.nack(msg, false, requeue);
        this.logger.debug('Message negatively acknowledged', {
          messageId,
          deliveryTag,
          requeue,
        });
      },
      onExtendDeadline: async (_seconds: number) => {
        // RabbitMQ doesn't support deadline extension
        this.logger.debug('Deadline extension not supported in RabbitMQ', { messageId });
      },
    });
  }

  /**
   * Get health status
   */
  async healthCheck(): Promise<HealthStatus> {
    const start = Date.now();

    try {
      if (!this._connected || !this.channel) {
        return {
          healthy: false,
          connected: false,
          error: 'Not connected',
        };
      }

      // Check queue exists
      await this.channel.checkQueue(this.rabbitConfig.queue.name);

      return {
        healthy: true,
        connected: true,
        latencyMs: Date.now() - start,
        details: {
          queue: this.rabbitConfig.queue.name,
          exchange: this.rabbitConfig.exchange?.name,
          paused: this._paused,
        },
      };
    } catch (error) {
      return {
        healthy: false,
        connected: false,
        latencyMs: Date.now() - start,
        error: error instanceof Error ? error.message : String(error),
      };
    }
  }
}

/**
 * Create a RabbitMQ consumer instance
 */
export function createRabbitMQConsumer<T = unknown>(
  config: RabbitMQConsumerConfig
): RabbitMQConsumer<T> {
  return new RabbitMQConsumer<T>(config);
}

File: go/rabbitmq/consumer.go
package rabbitmq

import (
	"context"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"

	"github.com/sns45/anyq/go/core"
)

// Consumer consumes messages from a RabbitMQ queue. It embeds *core.BaseConsumer
// and wires handler failures into ApplyStrategy, preserving the TypeScript
// legacy catch behaviour in the !handled branch.
//
// Capability notes (matching the shipped TS adapter):
//   - SupportsNativeDelay is false: a Park decision downgrades to in-process
//     retry (the base handles this).
//   - DeadLetterMessage is not overridden: the base default warns and nacks
//     without requeue, which lets the queue's x-dead-letter-exchange (when
//     configured via QueueConfig.DeadLetterExchange) route the message.
type Consumer struct {
	*core.BaseConsumer
	cfg ConsumerConfig

	conn        *amqp.Connection
	channel     *amqp.Channel
	consumerTag string
}

// NewConsumer builds a RabbitMQ consumer.
func NewConsumer(cfg ConsumerConfig) *Consumer {
	cfg.applyDefaults()
	bc := &core.BaseConsumer{}
	bc.InitBase(cfg.BaseQueueConfig)
	c := &Consumer{BaseConsumer: bc, cfg: cfg}
	bc.Bind(c)
	return c
}

// Connect dials the broker, opens a channel, sets QoS, declares the queue, and
// optionally declares and binds the configured exchange.
func (c *Consumer) Connect(ctx context.Context) error {
	if c.IsConnected() {
		return nil
	}
	if err := c.VerifyParkPolicy(); err != nil {
		return err
	}

	conn, err := amqp.DialConfig(dialURL(c.cfg.Connection), amqp.Config{
		Heartbeat:  c.cfg.Connection.Heartbeat,
		Dial:       amqp.DefaultDial(c.cfg.Connection.ConnectionTimeout),
		ChannelMax: uint16(c.cfg.Connection.ChannelMax),
		FrameSize:  c.cfg.Connection.FrameMax,
	})
	if err != nil {
		return core.NewConnectionError("failed to connect to RabbitMQ", err)
	}

	ch, err := conn.Channel()
	if err != nil {
		_ = conn.Close()
		return core.NewConnectionError("failed to open RabbitMQ channel", err)
	}

	if err := ch.Qos(c.cfg.Consumer.Prefetch, 0, false); err != nil {
		_ = ch.Close()
		_ = conn.Close()
		return core.NewConnectionError("failed to set prefetch", err)
	}

	args := c.queueArgs()
	if _, err := ch.QueueDeclare(
		c.cfg.Queue.Name,
		c.cfg.Queue.Durable,
		c.cfg.Queue.AutoDelete,
		c.cfg.Queue.Exclusive,
		false,
		args,
	); err != nil {
		_ = ch.Close()
		_ = conn.Close()
		return core.NewConnectionError("failed to declare queue", err)
	}

	if c.cfg.Exchange != nil {
		if err := ch.ExchangeDeclare(
			c.cfg.Exchange.Name,
			c.cfg.Exchange.Type,
			c.cfg.Exchange.Durable,
			c.cfg.Exchange.AutoDelete,
			c.cfg.Exchange.Internal,
			false,
			amqp.Table(c.cfg.Exchange.Arguments),
		); err != nil {
			_ = ch.Close()
			_ = conn.Close()
			return core.NewConnectionError("failed to declare exchange", err)
		}
		if err := ch.QueueBind(c.cfg.Queue.Name, c.cfg.BindingKey, c.cfg.Exchange.Name, false, nil); err != nil {
			_ = ch.Close()
			_ = conn.Close()
			return core.NewConnectionError("failed to bind queue", err)
		}
	}

	c.conn = conn
	c.channel = ch
	c.SetConnected(true)
	exchangeName := ""
	if c.cfg.Exchange != nil {
		exchangeName = c.cfg.Exchange.Name
	}
	c.Logger.Info("RabbitMQ consumer connected", map[string]any{
		"queue": c.cfg.Queue.Name, "exchange": exchangeName,
	})
	return nil
}

func (c *Consumer) queueArgs() amqp.Table {
	args := amqp.Table{}
	q := c.cfg.Queue
	if q.DeadLetterExchange != "" {
		args["x-dead-letter-exchange"] = q.DeadLetterExchange
		if q.DeadLetterRoutingKey != "" {
			args["x-dead-letter-routing-key"] = q.DeadLetterRoutingKey
		}
	}
	if q.MessageTTL > 0 {
		args["x-message-ttl"] = int32(q.MessageTTL)
	}
	if q.MaxLength > 0 {
		args["x-max-length"] = int32(q.MaxLength)
	}
	if q.MaxLengthBytes > 0 {
		args["x-max-length-bytes"] = int32(q.MaxLengthBytes)
	}
	for k, v := range q.Arguments {
		args[k] = v
	}
	if len(args) == 0 {
		return nil
	}
	return args
}

// Disconnect cancels the consumer and closes the channel and connection.
func (c *Consumer) Disconnect(ctx context.Context) error {
	if !c.IsConnected() {
		return nil
	}
	if c.channel != nil && c.consumerTag != "" {
		_ = c.channel.Cancel(c.consumerTag, false)
		c.consumerTag = ""
	}
	if c.channel != nil {
		_ = c.channel.Close()
		c.channel = nil
	}
	if c.conn != nil {
		_ = c.conn.Close()
		c.conn = nil
	}
	c.SetConnected(false)
	c.Logger.Info("RabbitMQ consumer disconnected", nil)
	return nil
}

// Subscribe consumes deliveries with a per-message handler. It blocks until ctx
// is cancelled, the deliveries channel closes, or a strategy Fail decision
// returns a fatal error.
func (c *Consumer) Subscribe(ctx context.Context, handler core.Handler, opts *core.SubscribeOptions) error {
	if c.channel == nil {
		return core.NewConnectionError("consumer not connected", nil)
	}

	deliveries, err := c.channel.ConsumeWithContext(
		ctx,
		c.cfg.Queue.Name,
		c.cfg.Consumer.ConsumerTag,
		c.cfg.Consumer.NoAck,
		c.cfg.Consumer.Exclusive,
		c.cfg.Consumer.NoLocal,
		false,
		c.consumeArgs(),
	)
	if err != nil {
		return core.NewConsumeError("failed to start consuming", err)
	}
	autoAck := opts.AutoAckEnabled()
	c.Logger.Info("subscribed to queue", map[string]any{"queue": c.cfg.Queue.Name})

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case d, ok := <-deliveries:
			if !ok {
				return ctx.Err()
			}
			if c.IsPaused() {
				_ = d.Nack(false, true)
				continue
			}
			if err := c.handleOne(ctx, handler, d, autoAck); err != nil {
				return err
			}
		}
	}
}

func (c *Consumer) handleOne(ctx context.Context, handler core.Handler, d amqp.Delivery, autoAck bool) error {
	msg := c.wrap(d)
	herr := handler(ctx, msg)
	if herr == nil {
		if autoAck && !c.cfg.Consumer.NoAck {
			return msg.Ack(ctx)
		}
		return nil
	}

	reinvoke := func() error { return handler(ctx, msg) }
	handled, applyErr := c.ApplyStrategy(ctx, msg, herr, reinvoke)
	if applyErr != nil {
		return applyErr // fatal: Fail decision or context cancellation
	}
	if handled {
		return nil
	}

	// Legacy catch behaviour (preserved from the TS adapter): log the error and
	// nack with requeue so the broker (or its dead-letter-exchange) can retry.
	c.Logger.Error("error processing message", map[string]any{"messageId": msg.ID(), "error": herr.Error()})
	if c.Config.OnError != nil {
		c.Config.OnError(herr)
	}
	return msg.Nack(ctx, true)
}

// SubscribeBatch consumes deliveries and dispatches them to a batch handler. It
// fills batches up to BatchSize or until BatchTimeout elapses.
func (c *Consumer) SubscribeBatch(ctx context.Context, handler core.BatchHandler, opts *core.SubscribeOptions) error {
	if c.channel == nil {
		return core.NewConnectionError("consumer not connected", nil)
	}

	deliveries, err := c.channel.ConsumeWithContext(
		ctx,
		c.cfg.Queue.Name,
		c.cfg.Consumer.ConsumerTag,
		false, // always manual ack for batch
		c.cfg.Consumer.Exclusive,
		c.cfg.Consumer.NoLocal,
		false,
		c.consumeArgs(),
	)
	if err != nil {
		return core.NewConsumeError("failed to start consuming", err)
	}

	batchSize := 10
	batchTimeout := 5 * time.Second
	if opts != nil {
		if opts.BatchSize > 0 {
			batchSize = opts.BatchSize
		}
		if opts.BatchTimeout > 0 {
			batchTimeout = opts.BatchTimeout
		}
	}
	autoAck := opts.AutoAckEnabled()

	var batch []amqp.Delivery
	timer := time.NewTimer(batchTimeout)
	timer.Stop()
	defer timer.Stop()

	flush := func() error {
		if len(batch) == 0 {
			return nil
		}
		current := batch
		batch = nil
		timer.Stop()
		return c.handleBatch(ctx, handler, current, autoAck)
	}

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-timer.C:
			if err := flush(); err != nil {
				return err
			}
		case d, ok := <-deliveries:
			if !ok {
				return ctx.Err()
			}
			if c.IsPaused() {
				_ = d.Nack(false, true)
				continue
			}
			if len(batch) == 0 {
				timer.Reset(batchTimeout)
			}
			batch = append(batch, d)
			if len(batch) >= batchSize {
				if err := flush(); err != nil {
					return err
				}
			}
		}
	}
}

func (c *Consumer) handleBatch(ctx context.Context, handler core.BatchHandler, deliveries []amqp.Delivery, autoAck bool) error {
	msgs := make([]core.Message, len(deliveries))
	for i, d := range deliveries {
		msgs[i] = c.wrap(d)
	}

	herr := handler(ctx, msgs)
	if herr == nil {
		if autoAck {
			for _, m := range msgs {
				_ = m.Ack(ctx)
			}
		}
		return nil
	}

	anyUnhandled := false
	for _, m := range msgs {
		mm := m
		reinvoke := func() error { return handler(ctx, []core.Message{mm}) }
		handled, applyErr := c.ApplyStrategy(ctx, mm, herr, reinvoke)
		if applyErr != nil {
			return applyErr
		}
		if !handled {
			anyUnhandled = true
		}
	}
	if anyUnhandled {
		c.Logger.Error("error processing batch", map[string]any{"count": len(msgs), "error": herr.Error()})
		if c.Config.OnError != nil {
			c.Config.OnError(herr)
		}
		for _, m := range msgs {
			_ = m.Nack(ctx, true)
		}
	}
	return nil
}

func (c *Consumer) consumeArgs() amqp.Table {
	args := amqp.Table{}
	if c.cfg.Consumer.Priority != 0 {
		args["x-priority"] = int32(c.cfg.Consumer.Priority)
	}
	for k, v := range c.cfg.Consumer.Arguments {
		args[k] = v
	}
	if len(args) == 0 {
		return nil
	}
	return args
}

func (c *Consumer) wrap(d amqp.Delivery) core.Message {
	messageID := d.MessageId
	if messageID == "" {
		messageID = "rmq-" + itoa(int(d.DeliveryTag))
	}
	ts := d.Timestamp
	if ts.IsZero() {
		ts = time.Now()
	}
	// RabbitMQ does not expose a delivery-attempt counter; mirror the TS adapter
	// which reports 2 for redelivered messages and 1 otherwise.
	deliveryAttempt := 1
	if d.Redelivered {
		deliveryAttempt = 2
	}

	headers := core.MessageHeaders{}
	for k, v := range d.Headers {
		headers[k] = headerBytes(v)
	}

	return core.NewMessage(core.MessageParams{
		ID:              messageID,
		Body:            d.Body,
		Key:             d.RoutingKey,
		Headers:         headers,
		Timestamp:       ts,
		DeliveryAttempt: deliveryAttempt,
		Metadata: core.ProviderMetadata{
			Provider: core.DriverRabbitMQ,
			RabbitMQ: &core.RabbitMQMetadata{
				Exchange:    d.Exchange,
				RoutingKey:  d.RoutingKey,
				ConsumerTag: d.ConsumerTag,
				DeliveryTag: d.DeliveryTag,
				Redelivered: d.Redelivered,
			},
		},
		Raw: d,
		OnAck: func(context.Context) error {
			return d.Ack(false)
		},
		OnNack: func(_ context.Context, requeue bool) error {
			return d.Nack(false, requeue)
		},
	})
}

// HealthCheck reports consumer health.
func (c *Consumer) HealthCheck(ctx context.Context) (core.HealthStatus, error) {
	start := time.Now()
	if !c.IsConnected() || c.channel == nil {
		return core.HealthStatus{Healthy: false, Connected: false, Error: "Not connected"}, nil
	}
	exchangeName := ""
	if c.cfg.Exchange != nil {
		exchangeName = c.cfg.Exchange.Name
	}
	return core.HealthStatus{
		Healthy:   true,
		Connected: true,
		LatencyMs: time.Since(start).Milliseconds(),
		Details:   map[string]any{"queue": c.cfg.Queue.Name, "exchange": exchangeName, "paused": c.IsPaused()},
	}, nil
}

// headerBytes coerces an AMQP header value into bytes.
func headerBytes(v any) []byte {
	switch t := v.(type) {
	case []byte:
		return t
	case string:
		return []byte(t)
	case nil:
		return nil
	default:
		return []byte(stringify(t))
	}
}

func stringify(v any) string {
	switch t := v.(type) {
	case int:
		return itoa(t)
	case int32:
		return itoa(int(t))
	case int64:
		return itoa(int(t))
	case bool:
		if t {
			return "true"
		}
		return "false"
	default:
		return ""
	}
}

var (
	_ core.Consumer      = (*Consumer)(nil)
	_ core.ConsumerHooks = (*Consumer)(nil)
)
